pdbtools/pdb_addter.py

```diff
@@ -0,0 +1,247 @@
+    if len(slices) == 3:
+        start, end, step = slices
+    elif len(slices) == 2:
+        start, end = slices
+    elif len(slices) == 1:
+        if option.startswith(':'):
+            end = slices[0]
+        elif option.endswith(':'):
+            start = slices[0]
+
+    if start is None:
+        emsg = 'ERROR!! Please specify starting value: \'{}\'\n'
+        sys.stderr.write(emsg.format(start))
+        sys.exit(1)
+    elif not is_integer(start):
+        emsg = 'ERROR!! Starting value must be numerical: \'{}\'\n'
+        sys.stderr.write(emsg.format(start))
+        sys.exit(1)
+    elif not (-999 <= int(start) < 9999):
+        emsg = 'ERROR!! Starting value must be between -999 and 9998: \'{}\'\n'
+        sys.stderr.write(emsg.format(start))
+        sys.exit(1)
+
+    if end is None:
+        end = 10000  # residues cannot reach this value (max 4 char)
+    elif not is_integer(end):
+        emsg = 'ERROR!! End value must be numerical: \'{}\'\n'
+        sys.stderr.write(emsg.format(end))
+        sys.exit(1)
+    elif not (-999 <= int(end) < 9999):
+        emsg = 'ERROR!! End value must be between -999 and 9998: \'{}\'\n'
+        sys.stderr.write(emsg.format(end))
+        sys.exit(1)
+
+    if step is None:
+        step = 1
+    elif not is_integer(step):
+        emsg = 'ERROR!! Step value must be numerical: \'{}\'\n'
+        sys.stderr.write(emsg.format(step))
+        sys.exit(1)
+    elif int(step) <= 0:
+        emsg = 'ERROR!! Step value must be a positive number: \'{}\'\n'
+        sys.stderr.write(emsg.format(step))
+        sys.exit(1)
+
+    start, end, step = int(start), int(end), int(step)
+
+    if start >= end:
+        emsg = 'ERROR!! Start ({}) cannot be larger than end ({})\n'
+        sys.stderr.write(emsg.format(start, end))
+        sys.exit(1)
+
+    resrange = set(range(start, end + 1))
+    return (fh, resrange, step)
+
+
+def run(fhandle, residue_range, step):
+    """
+    Add TER records within the residue range at frequency identified by step.
+
+    This function is a generator.
+
+    Parameters
+    ----------
+    fhandle : a line-by-line iterator of the original PDB file.
+
+    residue_range : set, list, or tuple
+        The residues describing the range to consider.
+
+    step : int
+        The step at which to insert a TER record.
+
+    Yields
+    ------
+    str (line-by-line)
+        All lines with added TER lines designated by inputs.
+    """
+    prev_res = None
+    res_counter = 0
+    records = ('ATOM')
+    for line in fhandle:
+        if line.startswith(records):
+
+            res_id = line[21:26]  # include chain ID
+            if res_id != prev_res:
+
+                prev_res = res_id
+                res_counter += 1
+                if res_counter - min(residue_range) != 0 and (res_counter - min(residue_range)) % step == 0 and int(line[22:26]) in residue_range:
+                    yield "TER\n"
+
+        yield line
```

Comment:
The previous `yield 'TER\n'` creates a `TER` repetition if `TER` is already present. For example, run with `data/dummy.pdb`.

```
python pdbtools/pdb_addter.py -1:4:1 tests/data/dummy.pdb

ATOM      3  HB3 ALA B   7      35.847  44.180   3.181  1.00  0.00           H  
ATOM      3  C   ALA B   7      36.213  46.067   1.258  1.00  0.00           C  
ATOM      3  O   ALA B   7      36.287  47.046   2.028  1.00  0.00           O  
TER       3      ALA B   7
TER
ATOM      3  N   ASN A   1      22.066  40.557   0.420  1.00  0.00           N  
ATOM      3  H   ASN A   1      21.629  41.305  -0.098  1.00  0.00           H  
ATOM      3  H2  ASN A   1      23.236  40.798   0.369  1.00  0.00           H  
```

Also, I used the number `4` for `end`, so I was not expecting the `TER` to be there since the chain starts at `4`, I expected on TER was added to that chain.

Reply:
This is likely the same issue as in the below comment. You are going from residue 7 (out of `resrange`) to residue 1 (in `resrange`). My proposed solution below for this issue should be the same. 

I will also resolve the duplicate TER entries.